rasa/nlu/featurizers/sparse_featurizer/lexical_syntactic_featurizer.py

```diff
@@ -69,82 +87,283 @@ def required_components(cls) -> List[Type[Component]]:
+                )
+            )
+            if not_in_config:
+                rasa.shared.utils.io.raise_warning(
+                    f"A feature to index mapping has been loaded that does not match "
+                    f"the configured features. The given mapping configures "
+                    f" (position in window, feature_name): {not_in_config}. "
+                    f" These are not specified in the given config "
+                    f" {self._feature_config}. "
+                    f" Will continue without these features."
+                )
 
-    def _create_feature_to_idx_dict(
-        self, training_data: TrainingData
-    ) -> Dict[Text, Dict[Text, int]]:
-        """Create dictionary of all feature values.
+    def train(self, training_data: TrainingData,) -> Resource:
+        """Trains the featurizer.
 
-        Each feature key, defined in the component configuration, points to
-        different feature values and their indices in the overall resulting
-        feature vector.
+        Args:
+          training_data: the training data
         """
+        feature_to_idx_dict = self._create_feature_to_idx_dict(training_data)
+        self._set_feature_to_idx_dict(feature_to_idx_dict=feature_to_idx_dict)
+        if not self._feature_to_idx_dict:
+            rasa.shared.utils.io.raise_warning(
+                "No lexical syntactic features could be extracted from the training"
+                "data. In order for this component to work you need to define "
+                "`features` that can be found in the given training data."
+            )
+        self.persist()
+        return self._resource
 
-        # get all possible feature values
-        all_features = []
+    def _create_feature_to_idx_dict(
+        self, training_data: TrainingData
+    ) -> Dict[Tuple[int, Text], Dict[Hashable, int]]:
+        """Create a nested dictionary of all feature values.
+
+        Returns:
+           a nested mapping that maps from tuples of positions (in the window) and
+           supported feature names to "raw feature to index" mappings, i.e.
+           mappings that map the respective raw feature values to unique indices
+           (where `unique` means unique with respect to all indices in the
+           *nested* mapping)
+        """
+        # collect all raw feature values
+        feature_vocabulary: Dict[Text, Set[Hashable]] = dict()
         for example in training_data.training_examples:
-            tokens = example.get(TOKENS_NAMES[TEXT])
-            if tokens:
-                all_features.append(self._tokens_to_features(tokens))
+            tokens = example.get(TOKENS_NAMES[TEXT], [])
+            sentence_features = self._map_tokens_to_raw_features(tokens)
+            for token_features in sentence_features:
+                for position_and_feature_name, feature_value in token_features.items():
+                    feature_vocabulary.setdefault(position_and_feature_name, set()).add(
```

Comment:
also getting some type warnings by PyCharm here and in 289

Reply:
👍   forgot to adapt some key here, they're not tuples of window index and feature name